File: 2024/java/day2/EvalSafety.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.lang.String;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class EvalSafety {

    private static boolean check_line(List<Integer> levels, int lo, int hi) {
        boolean rule1 = monotonic_levels(levels);
        boolean rule2 = rule1 && safe_differences(levels, lo, hi);
        return rule2;
    }

    private static boolean monotonic_levels(List<Integer> levels) {
        boolean increasing = false;
        for (int i=1; i < levels.size(); i++) {
            if (i == 1) {
                increasing = levels.get(1) > levels.get(0);
            } else {
                if (increasing && levels.get(i) <= levels.get(i-1)) return false;
                if (!increasing && levels.get(i) >= levels.get(i-1)) return false;
            }
        }
        return true;
    }

    private static boolean safe_differences(List<Integer> levels, int lo, int hi) {
        for (int i=1; i < levels.size(); i++) {
            int diff = Math.abs(levels.get(i) - levels.get(i-1));
            if (diff < lo || diff > hi) return false;
        }
        return true;
    }

    public static void main(String[] args) {
        String fileName = "input.txt";
        int i = 0;
        int safeLines = 0;
        int safeLinesIfModified = 0;

        try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
            String line;
            while ((line = br.readLine()) != null) {
                i += 1;
                String[] parts = line.trim().split("\\s+"); // handle multiple spaces
                if (parts.length > 1) {
                    List<Integer> levels = Arrays.stream(parts)
                                .map(Integer::parseInt)
                                .collect(Collectors.toList());
                    boolean ans = check_line(levels, 1, 3);
                    if (!ans) {
                        // create a collection of sublists and check them for safety
                        List<List<Integer>> sublists = IntStream.range(0, levels.size())
                            .mapToObj(x -> {
                                List<Integer> sublist = new ArrayList<>(levels);
                                sublist.remove(x);
                                return sublist;
                            })
                            .collect(Collectors.toList());
                        for (List<Integer> sublist : sublists) {
                            boolean ans1 = check_line(sublist, 1, 3);
                            if (ans1) {
                                safeLinesIfModified += 1;
                                break;
                            }
                        }  
                    } else {
                        safeLines += 1;
                    }
                }
            }
            int totalSafe = safeLines + safeLinesIfModified;
            System.out.println(safeLines + " lines of safe levels (without modification)");
            System.out.println(totalSafe + " lines of safe levels (allowing 1 level to be disregarded)");
                                        
        } catch (RuntimeException e) {
            System.out.println("Error found in line: " + i + " -> " + e.getMessage());
        } catch (IOException e) {
            System.out.println("Problem reading file: " + fileName + "-> " + e.getMessage());
        }
    }  
}

File: 2024/cpp/day2/evalsafety.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <vector>

using namespace std;

bool check_line(vector<int>& levels, int lo, int hi) {
    // track monotonicity direction
    bool mono_known = false;
    bool increasing = false;

    for (int i = 1; i < levels.size(); i++) {
        int diff = levels[i] - levels[i -1];
        // check difference against bounds
        if (abs(diff) < lo || abs(diff) > hi) return false;

        // check monotonicity; handle unknown case
        if (!mono_known) {
            if (diff != 0) { 
                increasing = diff > 0;
                mono_known = true;
            }
        } else {
            if ((increasing && diff < 0) || (!increasing && diff > 0)) {
                return false;
            }
        }
    }
    return true;
}

int main() {
    // params
    int lo = 1;
    int hi = 3;

    // tracking
    int safe_lines = 0;
    int safe_lines_modified = 0;

    ifstream inputFile("input.txt");
    if (!inputFile) {
        cerr << "Error: Unable to open file input.txt" << endl;
        return 1;
    }
    string line;
    while (getline(inputFile, line)) {
        istringstream lineStream(line);
        vector<int> levels;
        int value;

        // parse an unknown number of integers from the current line
        // std::istringstream automatically skips whitespace
        while (lineStream >> value) {
            levels.push_back(value);
        }
        if (levels.size() > 1) {
            if (!check_line(levels, lo, hi)) {
                // create sublists and check them for safety
                for (size_t i = 0; i < levels.size(); ++i) {
                    vector<int> sublist;
                    for (size_t j = 0; j < levels.size(); j++) {
                        if (j != i) {
                            sublist.push_back(levels[j]);
                        }
                    }
                    if (check_line(sublist, lo, hi)) {
                        safe_lines_modified += 1;
                        break;
                    }
                }
            } else {
                safe_lines += 1;
            }
        }
    }
    int total_safe = safe_lines + safe_lines_modified;
    cout << safe_lines << " lines of safe levels (without modification" << endl;
    cout << total_safe << " lines of safe levels (allowing 1 level to be disregarded)" << endl;
}

/*
* > clang --version
* > clang++ -std=c++17 evalsafety.cpp -o exec/evalsafety
* > exec/evalsafety
*/